README.md
Ember D3.js Component
==================

A graphic component for ember.js using d3.js

Installation
------------------

    bower install https://github.com/MythX/ember-d3-component.git
    
Include d3.js and ember-d3.js

    <script src="bower_components/ember-d3-component/dist/ember-d3.js"></script>

Usage
------------------

    {{ draw-chart width=1000 height=300 dataBinding='content' }}
    
Content example :

    {
        xAxis: {
          format: 'date',
          dateFormat: 'DD/MM',
          origin: false
        },
        yAxis: [{
          name: 'y1',
          color: 'black',
          legend: 'Legend 1',
          position: 'left'
        }],
        charts: [{
          type: 'bar',
          color: '#428BCA',
          yAxis: 'y1',
          data: [{
            keyD: your_key,
            valD: your_value
          }],
        }]
    }


Properties
------------------

Helper :

    width: Integer (not required)
    height: Integer (required)

General :

    xAxis: Object (required)
    yAxis: Array (required)
    charts: Array (required)

xAxis :

    format: String (date or numeric) (required)
    dateFormat: String (required if format = 'date')
    origin: Boolean (required)
    rotateLegend: Boolean (not required)

yAxis :

    name: String (required)
    color: String (not required)
    legend: String (not required)
    max: Integer (not required)
    position: String (left or right) (required)
    ticks: Integer (not required)

Line-chart :

    type: 'line' (required)
    yAxis: String (required)
    data: Array (required)
    color: String (required)
    gradient: Object (not required)
    animation: Boolean (not required)
    interpolate: String (not required) ['monotone', 'basic']
    
Bar-chart :

    type: 'bar' (required)
    yAxis: String (required)
    data: Array (required)
    animation: Boolean (not required)
    color: String (not required) - For single color
    colors: Array (not required) - For multi-color, you need one color for one bar
    
Area-chart:

    type: 'area' (required)
    yAxis: String (required)
    data: Array (required)
    color: String (required)
    gradient: Object (not required)
    interpolate: String (not required) ['monotone', 'basic']

Data : 

    [{
      keyD: key,
      valD: val,
     }]

Gradient : 

    gradient: {
      stops: [
        {offset: "0%", color: "red"},
        {offset: "50%", color: "orange"},
        {offset: "80%", color: "orange"},
        {offset: "100%", color: "green"}
      ],
      max: 100
    }

CHANGELOG.md
## Changes between 0.3.0 and 0.3.1

* Add color on y axis
* Add tick options

## Changes between 0.2.8 and 0.3.0

* Move dist files on dist/css and dist/js
* Add gradient on Area and Line

## Changes between 0.2.7 and 0.2.8

* Fix bug with yMax

## Changes between 0.2.6 and 0.2.7

* Add class circle for points on line

## Changes between 0.2.5 and 0.2.6

* Add animation on bar chart
* Add max on yAxis

## Changes between 0.2.4 and 0.2.5

* Fix bug with area-chart

## Changes between 0.2.3 and 0.2.4

* Resize chart

## Changes between 0.2.2 and 0.2.3

* The chart is now resizable

## Changes between 0.2.1 and 0.2.2

* Fix bug #2 (problem width on Firefox)

## Changes between 0.2.0 and 0.2.1

* X Axis legend can rotate with `rotateLegend` parameter

## Changes between 0.1.1 and 0.2.0

* `0.2.0` supports a new chart type (area)
* Fix bug with dual Y axis

gulpfile.js
var gulp        = require('gulp'),
    jshint      = require('gulp-jshint'),
    concat      = require('gulp-concat'),
    less        = require('gulp-less'),
    watch       = require('gulp-watch'),
    livereload  = require('gulp-livereload'),
    handlebars  = require('gulp-ember-handlebars'),
    header      = require('gulp-header'),
    runSequence = require('run-sequence'),
    pkg         = require('./package.json');

var banner = [
'// ==========================================================================',
'// Project:   Ember D3 Component',
'// Version    v<%= pkg.version %>',
'// Copyright: © 2014 Antoine Moser',
'// License:   MIT (see LICENSE)',
'// ==========================================================================\n',
].join('\n');

var paths = {
  templates: 'lib/templates/**/*.hbs',
  scripts: [
    'lib/templates-top.js',
    'lib/components/*.js'
  ],
//  js: 'dist/js/*.js',
  styles: 'lib/styles/ember-d3.less',
  dist: 'dist/**/*.*'
};

gulp.task('templates', function() {
  gulp.src([paths.templates])
    .pipe(handlebars({
      outputType: 'browser',
      namespace: 'Ember.TEMPLATES'
    }))
    .pipe(concat('templates.js'))
    .pipe(gulp.dest('tmp'));
});

gulp.task('scripts', function() {
  gulp.src(paths.scripts)
    .pipe(concat('components.js'))
    .pipe(gulp.dest('tmp'));
});

gulp.task('release', function() {
  gulp.src([
      'tmp/components.js'
    ])
    .pipe(concat('ember-d3.js'))
    .pipe(header(banner, { pkg: pkg }))
    .pipe(gulp.dest('dist/js/'));
});

gulp.task('styles', function() {
  return gulp.src(paths.styles)
    .pipe(less())
    .pipe(gulp.dest('dist/css/'));
});

gulp.task('watch', function() {
  var server = livereload();
  var changed = function(file) {
    server.changed(file.path);
  };

  gulp.watch(paths.scripts, ['scripts', 'release']);
  gulp.watch(paths.templates, ['templates']);
  gulp.watch(paths.styles, ['styles']);
//  gulp.watch(paths.js).on('change', changed);
  gulp.watch(paths.dist).on('change', changed);
});

gulp.task('default', ['templates', 'scripts', 'styles', 'release']);


src/main.js
(function(root, undefined) {

  Ember.LineChartComponent = Ember.Component.extend({
    tagName: 'svg',
    margin: {
      top: 35,
      right: 35,
      bottom: 35,
      left: 35
    },

    didInsertElement: function() {
      this._super();
      this.draw();
    },

    draw: function() {
      var width = this.get('width') - this.margin.left - this.margin.right,
          height = this.get('height') - this.margin.top - this.margin.bottom,
          data = this.get('data');

      this.height = height;

      var formatX = this.get('formatXAxis');
      var color = this.get('color');
      var viewXAxis = this.get('xAxis');
      var viewYAxis = this.get('yAxis');
      var toolTip = this.get('tooltip');

      var formatedData = this.formatData(data, formatX, 'DD/MM');

      var x = this.createX(formatedData, formatX);
      var y = this.createY(formatedData);

    
      if(color === undefined) {
        color = "#428bca";
      }

      var line = this.drawLine(x, y);

      /* Create chart */
      this.chart = d3.select('#'+this.get('elementId'))
        .attr('id', 'chart')
        .attr('width', width + this.margin.left + this.margin.right)
        .attr('height', height + this.margin.top + this.margin.bottom)
        .append('svg:g')
        .attr('transform', 'translate(' + this.margin.left + ',' + this.margin.top + ')');

      this.chart.append('svg:clipPath')
        .attr('id', 'clip')
        .append('svg:rect')
        .attr('width', width + this.margin.left + this.margin.right)
        .attr('height', height + this.margin.top + this.margin.bottom);

      this.chart.append('svg:path')
        .attr('class', 'line')
        .attr('clip-path', 'url(#clip)')
        .attr('stroke', color)
        .attr("class", "line")
        .attr('d', line(formatedData));

      var points = this.drawPoints(formatedData, x, y);

      if(toolTip) {
        this.drawTooltip(points, color);
      }

      if(viewXAxis) {
        this.drawXAxis(x);
      }

      if(viewYAxis) {
        var legendY = this.get("legendY");
        this.drawYAxis(y, legendY);
      }
    },

    createX: function(data, type) {
      var x = null;
      if(type === 'date') {
        x = d3.scale.ordinal().rangePoints([0, this.width]).domain(data.map(function(d) { return d.keyD; }));
      } else {
        x = d3.scale.linear().range([0, this.width]).domain([0, data.length-1]);
      }

      return x;
    },

    createY: function(data) {
      return d3.scale.linear().range([this.height, 0]).domain([0, d3.max(data, function(d) { return d.valD; })+10]);
    },

    drawLine: function(x, y) {
      var line = d3.svg.line().y(function(d) { return y(d.valD); });
      line.x(function(d) { return x(d.keyD); });
      return line;
    },

    drawXAxis: function(x, legendX, height) {
      var xAxis = d3.svg.axis().scale(x).orient("bottom");

      this.chart.append('svg:g')
        .attr('class', 'x axis')
        .attr('transform', 'translate(0,'+this.height+')')
        .call(xAxis);

      return xAxis;
    },

    drawYAxis: function(y, legendY) {
      var yAxis = d3.svg.axis().scale(y).orient("left");

      this.chart.append('svg:g')
        .attr('class', 'y axis')
        .call(yAxis)
        .append("text")
        .attr("transform", "rotate(-90)")
        .attr("y", 6)
        .attr("dy", ".71em")
        .style("text-anchor", "end")
        .text(legendY);

      return yAxis;
    },

    drawPoints: function(data, x, y) {
      var points = this.chart.selectAll('.points')
        .data(data)
        .enter().append('svg:circle')
        .attr('stroke', 'black')
        .attr('fill', 'black')
        .attr('cx', function(d) { return x(d.keyD); })
        .attr('cy', function(d) { return y(d.valD); })
        .attr('r', 3);

      return points;
    },

    drawTooltip: function(points, color) {
      var div = d3.select('body').append('div')
        .attr('class', 'tooltip')
        .style('opacity', 0);

      points.on('mouseover', function(d) {
        div.transition()
          .duration(200)
          .style('background', color)
          .style('opacity', 0.9);

        div.html('<strong>'+d.keyD+'</strong> <br/>' + d.valD.toFixed(1))
          .style('left', (d3.event.pageX) + 'px')
          .style('top', (d3.event.pageY - 28) + 'px');
      })
      .on('mouseout', function(d) {
        div.transition()
          .duration(500)
          .style('opacity', 0);
      });
    },

    formatData: function(data, type, format) {
      var formatedData = [];

      if(type === 'date') {
        data.forEach(function(date) {
          var formatDay = moment(date.keyD).format(format);
          var day = Ember.Object.create({
            keyD: formatDay,
            valD: date.valD
          });

          formatedData.push(day);
        });
      }

      return formatedData;
    },

    timeFormatter: function(date) {
      return moment(date).format('DD/MM');
    }
  });

  Ember.Handlebars.helper('line-chart', Ember.LineChartComponent);
}(this));

lib/components/bar-chart.js
(function() {

  Ember.Chart.BarChartComponent = Ember.Component.extend({
    tagName: 'svg',
    margin: {
      top: 35,
      right: 35,
      bottom: 35,
      left: 35
    },

    didInsertElement: function() {
      this.draw();
    },

    draw: function() {
      var width = this.get('width') - this.margin.left - this.margin.right,
          height = this.get('height') - this.margin.top - this.margin.bottom,
          data = this.get('data');

      var viewXAxis = this.get('xAxis');
      var viewYAxis = this.get('yAxis');
      var formatX = this.get('formatXAxis');
      var legendY = this.get('legendY');

      this.height = height;
      this.width = width;

      var x = this.createX(data);
      var y = this.createY(data);

      /* Create chart */
      this.chart = d3.select('#'+this.get('elementId'))
        .attr('id', 'bar-chart')
        .attr('width', width + this.margin.left + this.margin.right)
        .attr('height', height + this.margin.top + this.margin.bottom)
        .append('svg:g')
        .attr('transform', 'translate(' + this.margin.left + ',' + this.margin.top + ')');

      this.chart.append('svg:g')
        .attr('class', 'rects');

      this.chart.select('.rects').selectAll('rect')
        .data(data)
        .enter().append('rect')
        .attr('class', 'bar')
        .attr('x', function(d) { return x(d.keyD); })
        .attr('y', function(d) { return y(d.valD); })
        .attr('width', x.rangeBand())
        .attr('height', function(d) { return height - y(d.valD); })
        .attr('fill', function(d) {
          if(d.colorD === undefined)
            return '#428bca';
          else
            return d.colorD;
        });

      if(viewXAxis) {
        this.drawXAxis(x, data, formatX);
      }

      if(viewYAxis) {
        this.drawYAxis(y, legendY);
      }
    },

    createX: function(data, type) {
      return d3.scale.ordinal().rangeRoundBands([0, this.width], 0.1).domain(data.map(function(d) { return d.keyD; }));
    },

    createY: function(data) {
      return d3.scale.linear().range([this.height, 0]).domain([0, d3.max(data, function(d) { return d.valD; })]);
    },

    drawXAxis: function(x, data, formatXAxis, legendX, height) {
      var xAxis = d3.svg.axis().scale(x).orient("bottom");

      if(formatXAxis === 'date') {
        xAxis.ticks(d3.time.days, 1).tickFormat(this.timeFormatter);
      } else {
        xAxis.ticks(data.length);
      }

      this.chart.append('svg:g')
        .attr('class', 'x axis')
        .attr('transform', 'translate(0,'+this.height+')')
        .call(xAxis);

      return xAxis;
    },

    drawYAxis: function(y, legendY) {
      var yAxis = d3.svg.axis().scale(y).orient("left");
      
      this.chart.append('svg:g')
        .attr('class', 'y axis')
        .call(yAxis)
        .append("text")
        .attr('y', -20)
        .attr('x', -20)
        .attr('dy', '.71em')
        .text(legendY);

      return yAxis;
    },

    timeFormatter: function(date) {
      return moment(date).format('DD/MM');
    }
  });

  Ember.Handlebars.helper('bar-chart', Ember.Chart.BarChartComponent);

}());

lib/components/chart.js
(function() {

Ember.Chart.ChartComponent = Ember.Component.extend({
  tagName: 'div',
  classNames: ['chart'],
  margin: {
    top: 35,
    right: 30,
    bottom: 40,
    left: 35
  },

  didInsertElement: function() {
    this._super();
    this.initialize();
    this.draw();
  },

  initialize: function() {
    var width;
    this.containerSelector = '#'+this.get('elementId') + '.chart';
    if(this.get('width') === undefined) {
      width = parseInt(d3.select(this.containerSelector).style('width'), 10);
    } else {
      width = this.get('width');
    }
    width     = this.width  =  width - this.margin.left - this.margin.right;
    var height    = this.height = this.get('height') - this.margin.top - this.margin.bottom;

    this.chart = d3.select(this.containerSelector)
      .append('svg:svg')
      .style('width', this.width + this.margin.left + this.margin.right + 'px')
      .style('height', this.height + this.margin.top + this.margin.bottom + 'px')
      .append('svg:g')
      .attr('transform', 'translate(' + this.margin.left + ',' + this.margin.top + ')');
    
    this.defs = this.chart.append('defs');
  },

  draw: function(redraw) {
    var params    = this.params = this.get('data');

    var charts        = params.charts;
    var data          = [];
    var colorBars     = [];
    var colorBar      = [];
    var types         = [];
    var formatedDatas = [];
    var barChartAnimation = false;
    var formatXAxis   = params.formatXAxis || 'numeric';
    var yAxis         = params.yAxis;
    var xAxis         = params.xAxis;
    var yAxisArray    = [];
    var yAxisBar      = '';

    var lineCharts    = [];
    var areaCharts    = [];
    var barChart      = [];
    
    // Search YAxis
    yAxis.forEach(function(axis) {
      yAxisArray.push({
        name: axis.name,
        color: axis.color,
        legend: axis.legend,
        max: axis.max || 0,
        y: '',
        orient: axis.position,
        ticks: axis.ticks,
        data: []
      });
    });

    charts.forEach(function(dataset) {
      data.push(dataset.data);
      types.push(dataset.type);
      if (dataset.yAxis !== undefined) {
        yAxisArray.forEach(function(yAxis) {
          if(yAxis.name === dataset.yAxis) {
            yAxis.data.push(dataset.data);
          }
        });
      } else {
        yAxisArray[0].data.push(dataset.data);
      }

      var formatedData = this.formatData(dataset.data, xAxis);
      var interpolate = dataset.interpolate ? dataset.interpolate : '';
      
      if (dataset.type === 'line') {

        lineCharts.push(Ember.Object.create({
          data: formatedData,
          color: dataset.color,
          gradient: dataset.gradient,
          animation: dataset.animation && !redraw,
          points: true,
          interpolate: interpolate,
          yAxis: dataset.yAxis
        }));
      } else if(dataset.type === 'bar') {
        yAxisBar = dataset.yAxis;
        if(dataset.color !== undefined) {
          colorBars.push(dataset.color);
        } else {
          colorBar = dataset.colors;
        }
        if(dataset.animation !== undefined && dataset.animation === true) {
          barChartAnimation = true;
        }
      } else if(dataset.type === 'area') {
        areaCharts.push(Ember.Object.create({
          data: formatedData,
          color: dataset.color,
          gradient: dataset.gradient,
          interpolate: interpolate,
          yAxis: dataset.yAxis,
          defined: (dataset.defined) ? dataset.defined : false
        }));
      }
    }, this);

    data.forEach(function(d) {
      formatedDatas.push(this.formatData(d, xAxis));
    }, this);

    // Format barchat data
    var barChartData = this.barChartData(formatedDatas, types);
    var x    = this.x = this.createX(formatedDatas[0], xAxis.format, xAxis.origin, barChartData);

    // Create YAxis
    yAxisArray.forEach(function (yAxis) {
      if(yAxis.max === 0) {
        yAxis.max = this.searchMaxY(yAxis.data);
      }
      yAxis.y = this.createY(yAxis.max);
      this.drawYAxis(yAxis.y, yAxis.legend, yAxis.orient, yAxis.color, yAxis.ticks);
    }.bind(this));

    this.drawXAxis(x, xAxis.rotateLegend);
    
    // Draw lines and bar
    if (barChartData.length > 1)
      this.drawBarChart(this.height, x, this.searchYAxis(yAxisBar, yAxisArray), barChartData, colorBars, false, barChartAnimation);
    else if (barChartData.length === 1) {
      if(colorBars.length === 0) {
        this.drawBarChart(this.height, x, this.searchYAxis(yAxisBar, yAxisArray), barChartData, colorBar, true, barChartAnimation);
      } else {
        this.drawBarChart(this.height, x, this.searchYAxis(yAxisBar, yAxisArray), barChartData, colorBars, false, barChartAnimation);
      }
    }

    lineCharts.forEach(function(l) {
      
      var y = this.searchYAxis(l.yAxis, yAxisArray);
      
      if (l.gradient) {
        this.defs.append("linearGradient")
          .attr("gradientUnits", "userSpaceOnUse")
          .attr("id", l.yAxis + "-gradient")
          .attr("x1", 0).attr("y1", y(0))
          .attr("x2", 0).attr("y2", y(l.gradient.max))
        .selectAll("stop")
          .data(l.gradient.stops)
        .enter().append("stop")
          .attr("offset", function(d) { return d.offset; })
          .attr("stop-color", function(d) { return d.color; });
      }
      
      var line = this.drawLine(x, y, l.data, l.interpolate, xAxis.origin);
      var path = this.drawLineOnSvg(line, l.data, l.color, l.gradient ? l.yAxis + "-gradient" : null);
      this.drawPoints(l.data, x, y, xAxis.origin, l.color);

      if(l.animation === true) {
        this.lineAnimation(path);
      }
    }, this);

    areaCharts.forEach(function(a) {
      
      var y = this.searchYAxis(a.yAxis, yAxisArray);
      
      if (a.gradient) {
        this.defs.append("linearGradient")
          .attr("gradientUnits", "userSpaceOnUse")
          .attr("id", a.yAxis + "-gradient")
          .attr("x1", 0).attr("y1", y(0))
          .attr("x2", 0).attr("y2", y(a.gradient.max))
        .selectAll("stop")
          .data(a.gradient.stops)
        .enter().append("stop")
          .attr("offset", function(d) { return d.offset; })
          .attr("stop-color", function(d) { return d.color; });
      }
      
      var area = this.drawArea(x, y);
      this.drawAreaOnSvg(area, a.data, a.color, a.gradient ? a.yAxis + "-gradient" : null);
      var line = this.drawLine(x, y, a.data, a.interpolate, xAxis.origin);
      this.drawLineOnSvg(line, a.data, a.color, a.gradient ? a.yAxis + "-gradient" : null);

    }, this);

    d3.select(window).on('resize.'+this.containerSelector, function() {
      Ember.run.debounce(this, function() {
        var newWidth = parseInt(d3.select(this.containerSelector).style('width'), 10) - this.margin.left - this.margin.right;
        if(newWidth != this.width) {
          this.width = newWidth;

          d3.select(d3.select(this.containerSelector + ' > svg')[0][0])
            .style('width', this.width + this.margin.left + this.margin.right + 'px');

          this.clearCharts();
          this.draw(true);
        }
      }, 150);
    }.bind(this));
  },

  createX: function(data, type, origin, barChartData) {
    var x = null;
    if(type === 'date') {
      if(origin) {
        x = d3.scale.ordinal().rangePoints([0, this.width]).domain(data.map(function(d) { return d.keyD; }));
      } else {
        x = d3.scale.ordinal().rangeRoundBands([0, this.width], 0.1).domain(data.map(function(d) { return d.keyD; }));
      }
    } else if(type === 'numeric') {
      if(barChartData.length) {
        x = d3.scale.ordinal().rangeRoundBands([0, this.width], 0.1).domain(data.map(function(d) { return d.keyD; }));
      } else {
        x = d3.scale.linear().range([0, this.width]).domain([0, data.length - 1]);
      }
    }

    return x;
  },

  createY: function(max) {
    return d3.scale.linear()
      .range([this.height, 0])
      .domain([0, (this.get("yMax")) ? this.get("yMax") : max]);
  },

  drawLine: function(x, y, data, interpolate, origin, area) {
    var line = d3.svg.line().interpolate(interpolate)
      .y(function(d) { return y(d.valD); });
    if(!origin) {
      line.x(function(d) { return x(d.keyD) + (x.rangeBand() / 2); });
    } else {
      line.x(function(d) { return x(d.keyD); });
    }

    if (area !== undefined && area.defined())
      line.defined(area.defined());

    return line;
  },

  drawLineOnSvg: function(line, formatedData, color, gradientId) {
    var path = this.chart.append('svg:path')
      .attr('class', 'line')
      .attr('clip-path', 'url(#clip)')
      .attr('d', line(formatedData));
    
    if (gradientId) {
      path.attr('stroke', 'url(#'+gradientId+')');
    } else {
      path.attr('stroke', color);
    }

    return path;
  },

  lineAnimation: function(path) {
    var totalLength = path.node().getTotalLength();
    path.attr("stroke-dasharray", totalLength+","+totalLength)
      .attr("stroke-dashoffset", totalLength)
      .transition()
      .duration(1000)
      .ease("linear-in-out")
      .attr("stroke-dashoffset", 0);
  },

  drawPoints: function(data, x, y, origin, color) {
    var points = this.chart.selectAll('.points')
      .data(data)
      .enter().append('svg:circle')
      .attr('class', 'circle')
      .attr('data-toggle', 'tooltip')
      .attr('title', function(d) { return d.keyD + " </br>" + d.valD; })
      .attr("data-html", true)
      .attr('stroke', color)
      .attr('cx', function(d) { return x(d.keyD) + (!origin ? (x.rangeBand() / 2) : 0); })
      .attr('cy', function(d) { return y(d.valD); })
      .attr('r', 3);
    return points;
  },

  drawBarChart: function(height, x, y, data, colors, multicolor, animation) {
    var x1 = d3.scale.ordinal();
    var bars = data.map(function(d, i) { return  'bar'+i ; });
    x1.domain(bars).rangeRoundBands([0, x.rangeBand()]);

    var days = data[0].map(function(day) {
      return day.keyD;
    });

    var getValueForIndex = function(index, chart) {
      return chart[index].valD;
    };

    var values = days.map(function(day, i) {
      return {
        keyD: day,
        valD: data.map(getValueForIndex.bind(this, i))
      };
    });

    var bar = this.chart.selectAll(".bar")
      .data(values)
      .enter().append("g")
        .attr("class", "g")
        .attr("transform", function(d) { return "translate(" + x(d.keyD) + ",0)"; });
   
    if (data.length > 1 || !multicolor) {
      bar.selectAll("rect")
        .data(function (d) { return d.valD; })
        .enter().append("rect")
          .attr('class', 'bar')
          .attr("width", x1.rangeBand())
          .attr("x", function(d, i) { return x1("bar"+i); })
          .attr("y", this.height)
          .attr("height", 0)
          .attr("fill", function(d, i) { return colors[i]; });
    } else {
      var j = -1;
      bar.selectAll("rect")
        .data(function (d) { return d.valD; })
        .enter().append("rect")
          .attr("width", x1.rangeBand())
          .attr("x", function(d, i) { return x1("bar"+i); })
          .attr("y", this.height)
          .attr("height", 0)
          .attr("fill", function(d, i) { j++; return colors[j]; });
    }

    if(animation === true) {
     bar.selectAll("rect")
        .data(function (d) { return d.valD; })
        .transition()
        .duration(1000)
        .attr('y', function(d) { return y(d); })
        .attr('height', function(d) { return height - y(d); });
    } else {
      bar.selectAll("rect")
        .data(function (d) { return d.valD; })
        .attr("y", function(d) { return y(d); })
        .attr("height", function(d) { return height - y(d); });
    }
  },

  drawArea: function(x, y, defined) {
    var area = d3.svg.area()
      .interpolate('monotone')
      .x(function(d) { return x(d.keyD) + (x.rangeBand() / 2); })
      .y0(this.height)
      .y1(function(d) { return y(d.valD); });

    if (defined) {
      area.defined(function(d) {
        var res = true;
        defined.forEach(function(def) {
          if(def === d.valD && res !== false)
            res = false;
        });

        return res;
      });
    }

    return area;
  },

  drawAreaOnSvg: function(area, formatedData, color, gradientId) {
    var path = this.chart.append('svg:path')
      .attr('class', 'area')
      .attr('d', area(formatedData));

    if (gradientId) {
      path.attr('fill', 'url(#'+gradientId+')');
    } else {
      path.attr('fill', color);
    }
    
    return path;
  },

  drawXAxis: function(x, rotate) {
    var xAxis = d3.svg.axis().scale(x).orient("bottom");

    var axis = this.chart.append('svg:g')
      .attr('class', 'x axis')
      .attr('transform', 'translate(0,'+this.height+')')
      .call(xAxis);
      
    if (rotate === true) {
      axis.selectAll("text")
      .style("text-anchor", "end")
      .attr("transform", function(d) {
        return "rotate(-65)";
      });
    }

    return xAxis;
  },

  drawYAxis: function(y, legendY, orient, color, ticks) {
    var yAxis = d3.svg.axis().scale(y).orient(orient);

    if (ticks)
      yAxis.ticks(ticks);

    if (orient === 'left') {
      this.chart.append('svg:g')
        .attr('class', 'y y-left axis')
        .attr('fill', color)
        .call(yAxis)
        .append("text")
        .attr('y', -20)
        .attr('x', -20)
        .attr("dy", ".71em")
        .text(legendY);
      this.chart.select('.y-left').selectAll('path, line').style('stroke', color);
    } else {
      this.chart.append('svg:g')
        .attr('class', 'y y-right axis')
        .attr('fill', color)
        .attr('transform', 'translate('+(this.width)  + ',0)')
        .call(yAxis)
        .append("text")
        .attr('y', -20)
        .attr('x', 20)
        .attr("dy", ".71em")
        .style("text-anchor", "end")
        .text(legendY);
      this.chart.select('.y-right').selectAll('path, line').style('stroke', color);
    }
    return yAxis;
  },

  searchMaxY: function(datas) {
    var maxY = 0;
    datas.forEach(function(data) {
      data.forEach(function(d) {
        if(d.valD > maxY)
          maxY = d.valD;
      });
    });
    return maxY;
  },

  searchYAxis: function(name, yAxisArray) {
    var i = 0;

    while (i < yAxisArray.length && yAxisArray[i  ].name !== name) {
      i++;
    }

    return i < yAxisArray.length ? yAxisArray[i].y : null;
  },

  barChartData: function(data, type) {
    var result = [];

    for(var i = 0; i < type.length; i++) {
      if(type[i] === "bar") {
        result.push(data[i]);
      }
    }
    return result;
  },

  formatData: function(data, xAxis) {
    var formatedData = [];

    if(xAxis.format === 'date') {
      if(xAxis.dateFormat !== undefined) {
        data.forEach(function(date) {
          var formatDay = moment(date.keyD).format(xAxis.dateFormat);
          var day = Ember.Object.create({
            keyD: formatDay,
            valD: date.valD
          });
          formatedData.push(day);
        });
      }
    } else {
      formatedData = data;
    }
    return formatedData;
  },

  clearCharts: function() {
    this.chart.selectAll('.x').remove();
    this.chart.selectAll('.y').remove();
    this.chart.selectAll('circle').remove();
    this.chart.selectAll('.line').remove();
    this.chart.selectAll('.area').remove();
    this.chart.selectAll('.bar').remove();
    this.chart.selectAll('.g').remove();
  },

  updateCharts: function() {
    this.clearCharts();
    this.draw();
  }.observes('data')

});

Ember.Handlebars.helper('draw-chart', Ember.Chart.ChartComponent);
  
}());